lib/core/util.js

```diff
@@ -436,6 +436,46 @@ function parseHeaders (headers, obj) {
   return obj
 }
 
+function normalizeHeaders (headers) {
+  const ret = {}
+  if (headers == null) {
+    // Do nothing...
+  } else if (Array.isArray(headers)) {
+    for (let i = 0; i < headers.length; i += 2) {
+      if (Array.isArray(headers[i]) && headers[i].length === 2) {
+        const key = headerNameToString(headers[i][0])
+        const val = ret[key]
+        if (val == null) {
+          ret[key] = String(headers[i][1])
+        } else if (typeof val === 'string') {
+          ret[key] = [val, String(headers[i][1])]
+        } else {
+          val.push(String(headers[i][1]))
+        }
+      } else {
+        const key = headerNameToString(headers[i])
+        const val = ret[key]
+        if (val == null) {
+          ret[key] = String(headers[i + 1])
+        } else if (typeof val === 'string') {
+          ret[key] = [val, String(headers[i + 1])]
+        } else {
+          val.push(String(headers[i + 1]))
+        }
+      }
+    }
+  } else if (typeof headers[Symbol.iterator] === 'function') {
+    for (const [key, val] of headers) {
+      ret[headerNameToString(key)] = Array.isArray(val) ? val.map(x => String(x)) : String(val)
+    }
+  } else {
+    for (const [key, val] of Object.entries(headers)) {
+      ret[headerNameToString(key)] = Array.isArray(val) ? val.map(x => String(x)) : String(val)
+    }
+  }
+  return ret
+}
```

Comment:
This logic seems very complex. why is it needed?

Reply:
To normalize different headers shapes?